wasmtime/_memory.py

```diff
@@ -62,6 +64,57 @@ def data_ptr(self, store: Storelike) -> "ctypes._Pointer[c_ubyte]":
         """
         return ffi.wasmtime_memory_data(store._context, byref(self._memory))
 
+    def read(
+            self,
+            store: Storelike,
+            start: typing.Optional[int] = 0,
+            stop: typing.Optional[int] = None) -> typing.Any:
+        """
+        provide fast way to read large memory slice similar to list[start:stop]
+        negative start, stop is allowed in a way similat to list slice mylist[-10:]
+        """
+        data_ptr = self.data_ptr(store)
+        size = self.data_len(store)
+        key = slice(start, stop, None)
+        start, stop, _ = key.indices(size)
+        val_size = stop - start
+        if val_size <= 0:
+            # return bytearray of size zero
+            return bytearray(0)
+        ptr_type = ctypes.c_ubyte * val_size
+        src_ptr = (ptr_type).from_address(ctypes.addressof(data_ptr.contents) + start)
+        return bytearray(src_ptr)
+
+    def write(
+            self,
+            store: Storelike,
+            value: typing.Any,
+            start: typing.Optional[int] = 0,
+            stop: typing.Optional[int] = None) -> typing.Any:
+        """
+        write into a possibly large slice of memory
+        negative start, stop is allowed in a way similat to list slice mylist[-10:]
+        """
+        data_ptr = self.data_ptr(store)
+        size = self.data_len(store)
+        key = slice(start, stop)
+        start, stop, _ = key.indices(size)
+        # value must be bytearray ex. cast bytes() to bytearray
+        if not isinstance(value, array.array) and not isinstance(value, bytearray):
+            # value = array.array('B', value)
+            value = bytearray(value)
+        val_size = len(value)
+        # key.indices(size) knows about size but not val_size
+        stop = start + min(stop - start, val_size)
+        # NOTE: we can use * 1, because we need pointer to the start only
```

Comment:
I think this `NOTE` is outdated?

Reply:
no, it's not, the following code will work totally fine, since the actual size is read by `memmove()` and the pointer is just the start

```python
        # NOTE: we can use * 1, because we need pointer to the start only
        # ptr_type = ctypes.c_ubyte * val_size
        ptr_type = ctypes.c_ubyte * 1
        src_ptr = (ptr_type).from_buffer(value)
        dst_ptr = (ptr_type).from_address(ctypes.addressof(data_ptr.contents) + start)
        ctypes.memmove(dst_ptr, src_ptr, val_size)
```

Reply:
The "NOTE" here disagrees with the code below. The NOTE makes me expect `c_ubyte * 1` when below it shows up as `* val_size`. I think that discrepancy should be fixed.

Reply:
I agree with @alexcrichton . Currently, the comment is somewhat confusing because it is not reflected in the code. It just educates the reader on something related to the code.
So I suggest to either remove the comment, or actually use the * 1 and explain in the comment why it works
I'd prefer the latter, e.g.:
```
# Recommended way to create an array type from data type: multiply with a positive integer
array_type = ctypes.c_ubyte * 1
...
```

Reply:
@j-blue-arz I've removed the comment, and explanation why it worked because `ctypes.memmove(dst_ptr, src_ptr, val_size)` takes address of type `ctypes.c_char_p` or `ctypes.c_void_p` (it does not read size from the array size but rather from the 3rd param of `memmove`)